### PR TITLE
Find non-matching parallax and diffuse textures

### PR DESCRIPTION
**Pull request by Sebastian-81**
Missing TODOs: command line parameter to specify the past and testing with MO2

Filtering of the warnings is not implemented yet I see it as a nice-to-have feature and should probably be implemented in a way that can handle wildcards and both mod names as well as BSAs

### CODE REVIEW

File: ParallaxGen/src/main.cpp [hunk 1]
```diff
@@ -224,8 +225,20 @@ void mainRunner(ParallaxGenCLIArgs &Args, const filesystem::path &ExePath) {
   auto VanillaBSAList = PGC.getVanillaBSAList();
 
   // Map files
-  PGD.mapFiles(PGC.getNIFBlocklist(), PGC.getManualTextureMaps(), VanillaBSAList, !Args.NoMapFromMeshes, !Args.NoMultithread,
-               Args.HighMem);
+  PGD.mapFiles(PGC.getNIFBlocklist(), PGC.getManualTextureMaps(), VanillaBSAList, !Args.NoMapFromMeshes,
+               !Args.NoMultithread, Args.HighMem);
+
+  // TODO get path as command line argument
```
Comment (Sebastian-81) [inline ParallaxGen/src/main.cpp:231]: TODO

File: ParallaxGenLib/src/ParallaxGenDirectory.cpp [hunk 2]
```diff
@@ -425,3 +426,71 @@ auto ParallaxGenDirectory::getMeshes() const -> const unordered_set<filesystem::
...
+      DiffuseMod = MD.GetModFromTexture(DiffuseModPath);
+    } else {
+      std::filesystem::path DiffuseBSAPath = FileMap.at(DiffuseRelPath).BSAFile->Path;
+      DiffuseMod = DiffuseBSAPath.filename().wstring();
+    }
+
+    if (isLooseFile(ParallaxRelPath)) {
+      std::filesystem::path ParallaxFullPath = getLooseFileFullPath(ParallaxRelPath);
+      ParallaxModPath = MD.FindModPath(ParallaxFullPath);
+      ParallaxMod = MD.GetModFromTexture(ParallaxModPath);
```
Comment (hakasapl) [inline ParallaxGenLib/src/ParallaxGenDirectory.cpp:476]: This is static
```suggestion
      ParallaxMod = ModManagerDirectory::GetModFromTexture(ParallaxModPath);
```

File: ParallaxGenLib/src/ParallaxGenDirectory.cpp [hunk 3]
```diff
@@ -425,3 +426,71 @@ auto ParallaxGenDirectory::getMeshes() const -> const unordered_set<filesystem::
...
+    // determine diffuse and parallax mods
+    std::filesystem::path DiffuseModPath{};
+    std::filesystem::path ParallaxModPath{};
+    std::wstring DiffuseMod{};
+    std::wstring ParallaxMod{};
+
+    if (isLooseFile(DiffuseRelPath)) {
+      std::filesystem::path DiffuseFullPath = getLooseFileFullPath(DiffuseRelPath);
+      DiffuseModPath = MD.FindModPath(DiffuseFullPath);
+      DiffuseMod = MD.GetModFromTexture(DiffuseModPath);
```
Comment (hakasapl) [inline ParallaxGenLib/src/ParallaxGenDirectory.cpp:467]: This is static
```suggestion
      ParallaxMod = ModManagerDirectory::GetModFromTexture(ParallaxModPath);
```

**Review by hakasapl — CHANGES_REQUESTED** [review]
Left some comments, also unit tests would be good for the new methods.

I have written some code for the CLI argument piece, when I get access to your branch I can push them

File: ParallaxGenLib/src/ModManagerDirectory.cpp [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+  std::filesystem::recursive_directory_iterator ModIterator(StagingDir);
+  for (auto &CurItem : ModIterator) {
+    std::filesystem::path Path = CurItem.path();
+    if (std::filesystem::is_regular_file(Path) &&
+        (BethesdaDirectory::pathEqualityIgnoreCase(Path.extension(), std::filesystem::path{L".dds"}))) {
+
+      // note: do not precompute crc32 since it is very expensive
+      Textures.push_back({Path, std::filesystem::file_size(Path)});
+    }
+  }
```
[the remark this answers is not in the document]
Comment (Sebastian-81) [reply ParallaxGenLib/src/ModManagerDirectory.cpp:24]: Do you have something in mind where we need it now ot can we do it later?

[the remark this answers is not in the document]
Comment (hakasapl) [reply ParallaxGenLib/src/ModManagerDirectory.cpp:24]: I see, yeah no we don't need to read BSAs again here, but I think we should read what the winning bsa is from BethesdaDirectory and find it somewhere here as well (maybe also find BSA files?). Basically I want to be able to give a warning if a mod's diffuse is in a BSA and parallax in a different mod as loose file or BSA

[the remark this answers is not in the document]
Comment (Sebastian-81) [reply ParallaxGenLib/src/ModManagerDirectory.cpp:24]: That's implemented already here: https://github.com/hakasapl/ParallaxGen/blob/e5b8eb14ebdab8c5981bf1dfc85fd127d7e53d77/ParallaxGenLib/src/ParallaxGenDirectory.cpp#L469

File: ParallaxGen/src/main.cpp
Comment (Sebastian-81) [inline ParallaxGen/src/main.cpp:1]: There should be a warning to the user if he does not provide that directory ParallaxGen can not find mismatching textures and he may have visual errors

**Comment by hakasapl** [comment]
Closing in favor of https://github.com/hakasapl/ParallaxGen/pull/32

**Pull request closed without merging**